nexusclient/client.go

```diff
@@ -0,0 +1,725 @@
+}
+
+// Cancel requests to cancel an asynchronous operation.
+//
+// Cancelation is asynchronous and may be not be respected by the operation's implementation.
+func (h *OperationHandle) Cancel(ctx context.Context, options CancelOptions) error {
+	if h.id == "" {
+		return ErrHandleForSyncOperation
+	}
+
+	return h.client.cancelOperation(ctx, cancelOperationRequest{
+		Operation:   h.operation,
+		OperationID: h.id,
+		Header:      options.Header,
+	})
+}
+
+type cancelOperationRequest struct {
+	Operation   string
+	OperationID string
+	Header      http.Header
+}
+
+func (c *Client) cancelOperation(ctx context.Context, request cancelOperationRequest) error {
+	url, err := c.joinURL(request.Operation, request.OperationID, "cancel")
+	if err != nil {
+		return err
+	}
+	httpReq, err := http.NewRequestWithContext(ctx, "POST", url.String(), nil)
+	if err != nil {
+		return err
+	}
+	if request.Header != nil {
+		httpReq.Header = request.Header.Clone()
+	}
+
+	httpReq.Header.Set(headerUserAgent, UserAgent)
+	response, err := c.Options.HTTPClient.Do(httpReq)
+	if err != nil {
+		return err
+	}
+	defer response.Body.Close()
+	if response.StatusCode != http.StatusAccepted {
+		return c.newUnexpectedResponseError(fmt.Sprintf("unexpected response status: %q", response.Status), response)
+	}
+	return nil
+}
+
+// Input for [nexusclient.Client.StartOperation].
+type StartOperationRequest struct {
+	// Name of the operation to start.
+	Operation string
+	// Callback URL to provide to the handle for receiving async operation completions. Optional.
+	CallbackURL string
+	// Request ID that may be used by the server handler to dedupe this start request.
+	// By default a v4 UUID will be generated by the client.
+	RequestID string
+	// Header to attach to the HTTP request. Optional.
+	Header http.Header
+	// Body of the operation request.
+	// If it is an [io.Closer], the body will be automatically closed by the client.
+	Body io.Reader
+}
+
+// NewJSONStartOperationRequest is shorthand for creating a StartOperationRequest with a JSON body.
+// Marhsals the provided value to JSON using [nexusapi.DefaultMarshaler].
+func NewJSONStartOperationRequest(operation string, v any) (request StartOperationRequest, err error) {
+	req := StartOperationRequest{}
+	var b []byte
+	b, err = nexusapi.DefaultMarshaler(v)
+	if err != nil {
+		return
+	}
+	req.Operation = operation
+	req.Header = http.Header{nexusapi.HeaderContentType: []string{nexusapi.ContentTypeJSON}}
+	req.Body = bytes.NewReader(b)
+	return
+}
+
+// StartOperation calls the configured Nexus endpoint to start an operation.
+// The operation may complete synchronously, delivering the result inline, or asynchronously returning a reference that
+// can be used via the returned handle to interact with the operation.
+//
+// Use the returned [OperationHandle] to retrieve the operation's result.
+func (c *Client) StartOperation(ctx context.Context, request StartOperationRequest) (*OperationHandle, error) {
```

Comment:
I see this is why you made operation handle mutable, so you can lazily set state. I still think it should be immutable, but we can expose a `SynchronousResponse *http.Response` or something. Or we can return a return three results here instead of two which is starting to look like the better option.

It's looking more and more like using a handle for a synchronous response is just confusing. You have to close the handle, etc. Making the caller be aware that a response can be a synchronous or asynchronous up front I think is fine. Here's what I'm thinking:

```go
func (c *Client) StartOperation(ctx context.Context, request StartOperationRequest) (syncSuccess *http.Response, asyncSuccess *OperationHandle, err error)

func (c *Client) ExecuteOperation(ctx context.Context, request ExecuteOperationRequest) (*http.Response, error)
```

The latter works with sync and async for those that just want response

Reply:
Great suggestion, you should see this reflected soon.